File: src/ros2_bridge/mappers/foxglove_msgs/points_annotation.rs
//! Typed mapper for `foxglove_msgs/msg/PointsAnnotation`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn points_annotation_to_bus(msg: ros_env::foxglove_msgs::msg::PointsAnnotation) -> crate::foxglove_msgs::msg::v1::PointsAnnotation {
    crate::foxglove_msgs::msg::v1::PointsAnnotation {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        r#type: msg.type_ as i32,
        points: msg.points.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::point2::point2_to_bus).collect(),
        outline_color: Some(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_bus(msg.outline_color)),
        outline_colors: msg.outline_colors.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_bus).collect(),
        fill_color: Some(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_bus(msg.fill_color)),
        thickness: msg.thickness,
        metadata: msg.metadata.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::key_value_pair::key_value_pair_to_bus).collect(),
    }
}

pub(crate) fn points_annotation_to_ros(bus: crate::foxglove_msgs::msg::v1::PointsAnnotation) -> ros_env::foxglove_msgs::msg::PointsAnnotation {
    ros_env::foxglove_msgs::msg::PointsAnnotation {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        type_: bus.r#type as i32,
        points: bus.points.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::point2::point2_to_ros).collect(),
        outline_color: crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_ros(bus.outline_color.unwrap_or_default()),
        outline_colors: bus.outline_colors.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_ros).collect(),
        fill_color: crate::ros2_bridge::mappers::foxglove_msgs::color::color_to_ros(bus.fill_color.unwrap_or_default()),
        thickness: bus.thickness,
        metadata: bus.metadata.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::key_value_pair::key_value_pair_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsPointsAnnotationMapper;

impl TypedTopicMapper for FoxgloveMsgsPointsAnnotationMapper {
    type Ros = ros_env::foxglove_msgs::msg::PointsAnnotation;
    type Bus = crate::foxglove_msgs::msg::v1::PointsAnnotation;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(points_annotation_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(points_annotation_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/points_annotation.py
"""Generated mapper for `foxglove_msgs/msg/PointsAnnotation`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.point2 import point2_to_bus, point2_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.color import color_to_bus, color_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.key_value_pair import key_value_pair_to_bus, key_value_pair_to_ros

def points_annotation_to_bus(msg):
    from robot_bus.foxglove_msgs.msg.v1 import PointsAnnotation as BusMsg

    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.type = int(msg.type)
    bus.points.extend([point2_to_bus(x) for x in msg.points])
    bus.outline_color.CopyFrom(color_to_bus(msg.outline_color))
    bus.outline_colors.extend([color_to_bus(x) for x in msg.outline_colors])
    bus.fill_color.CopyFrom(color_to_bus(msg.fill_color))
    bus.thickness = msg.thickness
    bus.metadata.extend([key_value_pair_to_bus(x) for x in msg.metadata])
    return bus


def points_annotation_to_ros(bus):
    from foxglove_msgs.msg import PointsAnnotation as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.type = int(bus.type)
    out.points = [point2_to_ros(x) for x in bus.points]
    out.outline_color = color_to_ros(bus.outline_color)
    out.outline_colors = [color_to_ros(x) for x in bus.outline_colors]
    out.fill_color = color_to_ros(bus.fill_color)
    out.thickness = bus.thickness
    out.metadata = [key_value_pair_to_ros(x) for x in bus.metadata]
    return out


class FoxgloveMsgsPointsAnnotationMapper:
    def type_name(self) -> str:
        return "foxglove_msgs/msg/PointsAnnotation"

    def ros_msg_type(self):
        from foxglove_msgs.msg import PointsAnnotation as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return points_annotation_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.foxglove_msgs.msg.v1 import PointsAnnotation as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return points_annotation_to_ros(bus)
